elementpath/xpath_nodes.py

```diff
@@ -136,480 +161,755 @@ class NamespaceNode(XPathNode):
+        self.elem = elem
+        self.parent = parent
+        self.position = position
+
+    def __repr__(self) -> str:
+        return '%s(elem=%r)' % (self.__class__.__name__, self.elem)
 
     @property
     def name(self) -> str:
-        return self.attribute.name
+        try:
+            # an lxml PI
+            return cast(str, self.elem.target)  # type: ignore[attr-defined]
+        except AttributeError:
+            return cast(str, self.elem.text).split(' ', maxsplit=1)[0]
 
-    def as_item(self) -> Tuple[str, Any]:
-        return self.attribute.name, self.value
+    @property
+    def value(self) -> ElementProtocol:
+        return self.elem
+
+    @property
+    def string_value(self) -> str:
+        return self.elem.text or ''
+
+    @property
+    def typed_value(self) -> str:
+        return self.elem.text or ''
+
+
+class ElementNode(XPathNode):
+    """
+    A class for processing XPath element nodes that uses lazy properties to
+    diminish the average load for a tree processing.
+    """
+    kind = 'element'
+    elem: Union[ElementProtocol, XsdElementProtocol, XMLSchemaProtocol]
+    nsmap: Dict[Optional[str], str]
+    xsd_type: Optional[XsdTypeProtocol]
+    _namespaces: Optional[List['NamespaceNode']]
+    attributes: Optional[List['AttributeNode']]
+    children: List[ChildNodeType]
+
+    __slots__ = 'nsmap', 'elem', 'parent', 'position', 'xsd_type', \
+                '_namespaces', 'attributes', 'children'
+
+    def __init__(self,
+                 elem: Union[ElementProtocol, XsdElementProtocol, XMLSchemaProtocol],
+                 parent: Optional[Union['ElementNode', 'DocumentNode']] = None,
+                 position: int = 1,
+                 nsmap: Optional[Any] = None,
+                 xsd_type: Optional[XsdTypeProtocol] = None) -> None:
+
+        self.elem = elem
+        self.parent = parent
+        self.position = position
+        self.nsmap = {} if nsmap is None else nsmap
+        self.xsd_type = xsd_type
+        self._namespaces = None
+        self.attributes = None
+        self.children = []
 
     def __repr__(self) -> str:
-        return '%s(name=%r)' % (self.__class__.__name__, self.attribute.name)
+        return '%s(elem=%r)' % (self.__class__.__name__, self.elem)
 
-    def __eq__(self, other: Any) -> bool:
-        return isinstance(other, self.__class__) and \
-            self.attribute == other.attribute and \
-            self.value == other.value
+    def is_schema_element(self) -> bool:
+        return hasattr(self.elem, 'name') and hasattr(self.elem, 'type')
 
-    def __hash__(self) -> int:
-        return hash((self.attribute, self.value))
+    @property
+    def namespaces(self) -> List['NamespaceNode']:
+        if self._namespaces is None:
+            # Lazy generation of namespace nodes of the element
+            position = self.position + 1
+            self._namespaces = [NamespaceNode('xml', XML_NAMESPACE, self, position)]
```

Comment:
It's usually worth using a local variable while building some non-trivial data structure, instead of incrementally updating globally visible state.

Reply:
Node builders reserve positions for namespace nodes, built lazy if required by the parser. The position of the element node is not changed.